### PR TITLE
Initial docker container

### PR DESCRIPTION
**Pull request by fredclausen**

### CODE REVIEW

**Comment by mikenye** [comment]
Hi @fredclausen, thanks so much for your hard work on this. I'll be adding comments as I review through your PR.

**Comment by mikenye** [comment]
Regarding your `.gitignore` file, to globally ignore `.DS_Store` files, you could just have a single line as follows:

```gitignore
.DS_Store
```

This way, git will ignore `.DS_Store` files in all directories.

Handy hint: You can use <http://gitignore.io> to generate .gitignore files based on your IDE, OS, environment, etc. If I use that tool and enter the keywords "macOS" and "vscode", I get the following:

```gitignore

# Created by https://www.toptal.com/developers/gitignore/api/macos,vscode
# Edit at https://www.toptal.com/developers/gitignore?templates=macos,vscode

### macOS ###
# General
.DS_Store
.AppleDouble
.LSOverride

# Icon must end with two \r
Icon


# Thumbnails
._*

# Files that might appear in the root of a volume
.DocumentRevisions-V100
.fseventsd
.Spotlight-V100
.TemporaryItems
.Trashes
.VolumeIcon.icns
.com.apple.timemachine.donotpresent

# Directories potentially created on remote AFP share
.AppleDB
.AppleDesktop
Network Trash Folder
Temporary Items
.apdisk

### vscode ###
.vscode/*
!.vscode/settings.json
!.vscode/tasks.json
!.vscode/launch.json
!.vscode/extensions.json
*.code-workspace

# End of https://www.toptal.com/developers/gitignore/api/macos,vscode
```

File: buildx.sh [hunk 1]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env sh
+#shellcheck shell=sh
+
+set -xe
+
+REPO=mikenye
+IMAGE=planefinder
```
Comment (mikenye) [inline buildx.sh]: I think this should be "adsbhub"

File: buildx.sh [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+PLATFORMS="linux/amd64,linux/arm/v7,linux/arm64"
```
Comment (mikenye) [inline buildx.sh:8]: I'll try building this on `linux/arm/v6` too, which we can add after the PR too.

File: rootfs/etc/cont-init.d/01-adsbhubclient [hunk 3]
```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/with-contenv sh
+# shellcheck shell=sh
+
+# Check to make sure the correct command line arguments have been set
+EXITCODE=0
+if [ -z "${BEASTPORT}" ]; then
+  echo "BEASTPORT environment variable not set. Using default"
+  export BEASTPORT=30002
```
Comment (mikenye) [inline rootfs/etc/cont-init.d/01-adsbhubclient]: This might not work the way you intend. Exporting here will only be valid for the execution of this script. I can see you reference it in `rootfs/usr/bin/adsbhub.sh`, which will use `30005`; the defined value from Dockerfile. I'd probably remove the export line so all the defaults are defined in the Dockerfile.

File: rootfs/etc/cont-init.d/01-adsbhubclient [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+fi
+
+# Set up timezone
+if [ -z "${TZ}" ]; then
+  echo "WARNING: TZ environment variable not set"
+else
+  ln -snf "/usr/share/zoneinfo/$TZ" /etc/localtime && echo "$TZ" >/etc/timezone
+fi
+
+chmod +x /usr/bin/adsbhub.sh
```
Comment (mikenye) [inline rootfs/etc/cont-init.d/01-adsbhubclient]: This is a perfectly valid way of ensuring that the script is executable and I have no problems with this.

If we wanted to, we could leverage s6-overlay to do this (and keep all the permissions/modes kept in one place). See: <https://github.com/just-containers/s6-overlay#fixing-ownership--permissions>

File: rootfs/usr/bin/adsbhub.sh [hunk 5]
```diff
@@ -0,0 +1,66 @@
+#!/bin/bash
+# ------------------------------------------------------------------
+# www.adsbhub.org
+# version: 1.06
+# ------------------------------------------------------------------
+
+ckey="${CLIENTKEY}"
+#cmd="nc -w 60 -q 10 ${BEASTHOST} ${BEASTPORT} | nc -w 60 -q 10 data.adsbhub.org 5001"
+cmd="socat TCP4:${BEASTHOST}:${BEASTPORT} TCP4:data.adsbhub.org:5001"
+#cmd="nc -w 60 -q 10 localhost 30002 | nc -w 60 -q 10 94.130.23.233 5001"
```
Comment (mikenye) [inline rootfs/usr/bin/adsbhub.sh:10]: Original command has:

* `-w 60` - connection timeout after 60 seconds
* `-q 10` - sleep 10 seconds after connection ends

Presumably these are there for a reason, so if we're going to use `socat`, we should try and do the same thing.

I _think_ we want `-t 10 -T 60` added to the `socat` command (before the existing arguments).

Also, for security, we probably want to leverage `socat`'s unidirectional mode, so I'd also add `-u`, that way data can only go from local to ADSBHub (and not the other direction).

File: rootfs/etc/services.d/logwatcher/run [hunk 6]
```diff
@@ -0,0 +1,5 @@
+#!/usr/bin/with-contenv bash
```
Comment (mikenye) [inline rootfs/etc/services.d/logwatcher/run]: I don't think we need this file - looking at the adsbhub script, all logging will be to stdout/stderr.

File: rootfs/healthcheck.sh [hunk 7]
```diff
@@ -0,0 +1,33 @@
+#!/usr/bin/env bash
+
+set -e
+
+# Get latest log file
+PFCLIENT_LOG_FILE=$(find /var/log/pfclient -type f -iname "pfclient-log_*.log" | sort | tail -1)
```
Comment (mikenye) [inline rootfs/healthcheck.sh]: It looks like this healthcheck will need to be re-written for ADSBHub. I'd suggest removing it for now and we can tackle writing a healthcheck once the container is working.

File: README.md [hunk 8]
```diff
@@ -1,98 +1,143 @@
...
+
+```shell
+docker run \
+ -d \
+ --rm \
+ --name adsbhub \
+ -e TZ="YOURTIMEZONE" \
+ -e BEASTHOST=YOURBEASTHOST \
+ -e CLIENTKEY=YOURSHARECODE \
+ mikenye/adsb-hub
```
Comment (mikenye) [inline README.md]: I think the repo name should be `mikenye/adsbhub` (no hyphen).

File: README.md [hunk 9]
```diff
@@ -1,98 +1,143 @@
...
+    restart: always
+    devices:
+      - /dev/bus/usb/001/007:/dev/bus/usb/001/007
+    networks:
+      - adsbnet
+    command:
+      - --dcfilter
+      - --device-type=rtlsdr
+      - --fix
+      - --forward-mlat
```
Comment (mikenye) [inline README.md]: `--forward-mlat` should be removed

File: README.md [hunk 10]
```diff
@@ -1,98 +1,143 @@
...
+      - --metric
+      - --mlat
+      - --modeac
+      - --ppm=0
+      - --net
+      - --stats-every=3600
+      - --quiet
+      - --write-json=/var/run/readsb
+
+  pfclient:
```
Comment (mikenye) [inline README.md]: Need to replace with adsbhub.

File: README.md [hunk 11]
```diff
@@ -1,98 +1,143 @@
...
+## Ports
+
+No ports are exposed in this container
+
+## Environment variables
+
+| Variable | Description | Default |
+|----------|-------------|---------|
+| `TZ` | Timezone for the container. | Unset|
+| `BEASTHOST` | Host for RAW ADSB packets. Required. | Unset |
```
Comment (mikenye) [inline README.md]: Looking at ADSBHub's "how to feed" page, it says they support data in "raw Beast hex format or SBS/30003". 

Something to consider, should we also have a `BASESTATIONHOST`/`PORT` or make the variable name more generic? `DATAHOST`/`PORT`?

File: Dockerfile [hunk 12]
```diff
@@ -0,0 +1,32 @@
+FROM debian:stable-slim
+
+ENV BEASTPORT=30005 \
+    S6_BEHAVIOUR_IF_STAGE2_FAILS=2
+
+RUN set -x && \
+   apt-get update && \
+    apt-get install --no-install-recommends -y \
```
Comment (mikenye) [inline Dockerfile]: We can probably reduce what we install (and keep the image size down. 

Looking at `rootfs/usr/bin/adsbhub.sh`, we'll need:

* ca-certificates
* wget (can likely swap curl out for this)
* net-tools (for netstat)
* socat

And temporarily (during container build), we'll need:

* gnupg (to verify s6-overlay)
* file (so the s6-overlay installer script can determine the correct architecture)

Have a look at how I've started managing this here: <https://github.com/mikenye/docker-readsb-protobuf/blob/54edf941e71bcb9aaaca63cbe35d0bf9d4d69b7b/Dockerfile#L65>

The `KEPT_PACKAGES` packages are installed. The `TEMP_PACKAGES` packages are removed curing the clean-up at the end of the container build: <https://github.com/mikenye/docker-readsb-protobuf/blob/54edf941e71bcb9aaaca63cbe35d0bf9d4d69b7b/Dockerfile#L258>

**Review by mikenye — CHANGES_REQUESTED** [review]
OK @fredclausen this is a fantastic effort and I congratulate you on your first PR. I've made a whole bunch of review comments that I'd like you to take a look at prior to merging this, if that's OK.

Apologies if my review notes are too wordy, my goal was to try and educate.

Happy to discuss here or in our Discord.

File: buildx.sh (hunk 1 above)
Comment (fredclausen) [reply buildx.sh]: fixed

Comment (mikenye) [reply buildx.sh]: Nice one :)

**Review by fredclausen — COMMENTED** [review]
Made the requested changes

**Review by fredclausen — COMMENTED** [review]
Fixed issues

**Review by mikenye — APPROVED** [review]
I made some minor changes, great job.
